### PR TITLE
Include `underscore.sty`

### PR DESCRIPTION
**Pull request by kir0ul**
## PR Summary
This PR closes https://github.com/matplotlib/matplotlib/issues/17774.
It includes the `underscore.sty` package in the LaTeX preamble to take care of underscores.
Most of the changes were proposed by @anntzer in https://github.com/matplotlib/matplotlib/pull/18357#issuecomment-683410662.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

### CODE REVIEW

**Comment by jklymak** [comment]
Looks like you based this off an old version of master.  Can you update?

**Comment by jklymak** [comment]
(ooops sorry for hitting the close button)

**Comment by jklymak** [comment]
I'm not sure vendoring `underscore.sty` is a great solution, and I'm not sure what the consequences of including our vendored version in everyone's preamble are.  Is there a reason we can't just suggest people install underscore.sty themselves, and include it in their preamble?  Otherwise, I stand by my assertion this is a pandas problem.

**Comment by jklymak** [comment]
Sure, you can reproduce w/o pandas, but it is pandas that is inserting this automatically and causing the headache.  Presumably most folks who want to use LaTeX for anything know not to put underscores anywhere.  

As for "officially unsupported" I'm not suggesting we start support mucking with the preamble, just that it is a viable workaround for people who have this problem.  I'm not sure the alternative of forcing the use of `underscore.sty` doesn't have other un-intended consequences.  I assume it is an optional style file for good reason.

File: lib/matplotlib/tests/test_texmanager.py [hunk 1]
```diff
@@ -42,3 +42,12 @@ def test_font_selection(rc, preamble, family):
     src = Path(tm.make_tex("hello, world", fontsize=12)).read_text()
     assert preamble in src
     assert [*re.findall(r"\\\w+family", src)] == [family]
+
+def test_usetex_with_underscore():
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_texmanager.py:46]: This probably should go in `test_usetex.py` which already has a guard at the top to skip the tests when latex is not available.

**Comment by tacaswell** [comment]
I am weakly in favor of this as it brings the allowed strings when using latex or not in sync.   Because our internal mathtex engine only kicks in when in math mode (e.g. `$XYZ$`) and not for normal text which means that:

```python
fig, ax0 = plt.subplots()
ax0.text(.5, .5, 'a\_b')
ax0.text(.5, .4, 'a\_b', usetex=True)
fig.savefig('/tmp/so.png')
plt.show()

```

so

renders significantly differently if you are using latex or not.   On further consideration I think that my suggestion in https://github.com/matplotlib/matplotlib/issues/17774#issuecomment-652681674 is wrong and that we can not pre-process strings as we have no way of knowing what the usetex state will be when we actually get to render time.

I think the options here are : 

1. make underscore.sty a required latex dependency (pro: easy on us con: might be hard on the users)
2. make underscore.sty an optional latex dependency  (pro: easy on us, won't explode on users, con: leaves behavior silently dependent on installation of a latex package)
3. vendor and inject underscore.sty as this PR is doing
4. write a Python side function the at draw time in `Text.draw` to try and escape any `_` not in math mode

I'm inclined to 3 or 4.

----

Pandas is not at fault here, they are just an abundant source of labels with `_` in them that expose the underlying inconsistency in how text `_` are handled by us vs LaTeX.

**Comment by kir0ul** [comment]
Refactored this PR according to [what has been decided during the meeting](https://hackmd.io/HYCHlmH7SdarZWkvVucF-g?view#underscores-in-math-mode).
Let me know if I missed anything :slightly_smiling_face:

**Comment by timhoffm** [comment]
There are two TeX-related test failures. Can you make sense of these?

**Comment by kir0ul** [comment]
Hum... not completely sure what's happening here since I can't reproduce the tests failing on my local machine...

However, having a closer look at the `underscore` package docs, they say:
> **Deficiencies**
[...]
You must avoid “_” in file names and in cite or ref tags, or you must use the babel package,
with its active-character controls, or you must give the [strings] option, which attempts to
redefine several commands (and may not work perfectly). Even without the [strings] option
or babel, you can use occasional underscores like: “\include{file\string_name}”.

And since all the failing tests have an underscore in the file name like in the error below, I'd suspect this is the problem:
```
! Missing number, treated as zero.
<to be read again> 
                   \protect 
l.36 ...aphics*[angle=90]{/tmp/tmp_fyboc3t/tmp.ps}
                                                  
No pages of output.
```

**Comment by jklymak** [comment]
Looks like you fixed it, except for a listing error ;-)

**Comment by anntzer** [comment]
I think(?) the additional dependency needs an api_changes entry?

**Comment by QuLogic** [comment]
Yes, that would be preferred; it should go in the `development` section.

File: lib/matplotlib/tests/test_usetex.py [hunk 2]
```diff
@@ -100,8 +100,22 @@ def test_usetex_packages(pkg):
...
 
 
-def test_textcomp_full():
-    plt.rcParams["text.latex.preamble"] = r"\usepackage[full]{textcomp}"
+@pytest.mark.parametrize(
+    "latex_statement",
+    [r"\usepackage[full]{textcomp}", r"\usepackage{underscore}"],
+)
+def test_latex_pkg_already_loaded(latex_statement):
+    plt.rcParams["text.latex.preamble"] = latex_statement
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_usetex.py:108]: I'd just name this `preamble`:
```suggestion
@pytest.mark.parametrize(
    "preamble",
    [r"\usepackage[full]{textcomp}", r"\usepackage{underscore}"],
)
def test_latex_pkg_already_loaded(preamble):
    plt.rcParams["text.latex.preamble"] = preamble
```

**Review by QuLogic — APPROVED** [review]

**Review by jklymak — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @kir0ul! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**